### ISSUE TITLE
Overlapping Gateway HTTPS hosts causes ACK ERROR, rejecting full listener

### ISSUE THREAD

**Issue by howardjohn** [istio/istio#13717]
Pilot gives 
```
ADS:LDS: ACK ERROR 10.60.53.62:48764 istio-ingressgateway-9c5dfc8c6-h6fbg.istio-system node:<id:"router~10.60.53.62~istio-ingressgateway-9c5dfc8c6-h6fbg.istio-system~istio-system.svc.cluster.local" cluster:"istio-ingressgateway" metadata:<fields:<key:"CONFIG_NAMESPACE" value:<string_value:"istio-system" > > fields:<key:"ISTIO_META_INSTANCE_IPS" value:<string_value:"10.60.53.62,10.60.53.62,fe80::40fa:beff:fe06:8582" > > fields:<key:"ISTIO_PROXY_SHA" value:<string_value:"istio-proxy:a169a0c0cd86b51538c240e2d037fa8f7f5860ae" > > fields:<key:"ISTIO_PROXY_VERSION" value:<string_value:"1.1.3" > > fields:<key:"ISTIO_VERSION" value:<string_value:"1.1.3" > > fields:<key:"POD_NAME" value:<string_value:"istio-ingressgateway-9c5dfc8c6-h6fbg" > > fields:<key:"ROUTER_MODE" value:<string_value:"sni-dnat" > > fields:<key:"USER_SDS" value:<string_value:"true" > > fields:<key:"istio" value:<string_value:"sidecar" > > > locality:<> build_version:"a169a0c0cd86b51538c240e2d037fa8f7f5860ae/1.11.0-dev/Clean/RELEASE/BoringSSL" > type_url:"type.googleapis.com/envoy.api.v2.Listener" response_nonce:"c0ee59dc-db9c-497d-91b0-e859ad5462d9" error_detail:<code:13 message:"Error adding/updating listener(s) 0.0.0.0_443: error adding listener '0.0.0.0:443': multiple filter chains with effectively equivalent matching rules are defined" >
```

Relevant part is `multiple filter chains with effectively equivalent matching rules are defined`.

This causes the full 0.0.0.0_443 listener to be rejected, effectively breaking all ingress traffic.

Reproduce with:
Follow steps in https://istio.io/docs/tasks/traffic-management/secure-ingress/mount/ to create secret

Make Gateway config:
```
apiVersion: networking.istio.io/v1alpha3
kind: VirtualService
metadata:
  name: httpbin
spec:
  hosts:
  - httpbin.example.com
  gateways:
  - mygateway
  tls:
  - match:
    - port: 443
      sni_hosts:
      - httpbin.example.com
    route:
    - destination:
        host: my-httpbin
        port:
          number: 443
---
apiVersion: networking.istio.io/v1alpha3
kind: Gateway
metadata:
  name: mygateway
spec:
  selector:
    istio: ingressgateway # use istio default ingress gateway
  servers:
  - port:
      number: 443
      name: https
      protocol: HTTPS
    tls:
      mode: SIMPLE
      serverCertificate: /etc/istio/ingressgateway-certs/tls.crt
      privateKey: /etc/istio/ingressgateway-certs/tls.key
    hosts:
    - httpbin.example.com

  - port:
      number: 443
      name: https-other
      protocol: HTTPS
    tls:
      mode: SIMPLE
      serverCertificate: /etc/istio/ingressgateway-certs/tls.crt
      privateKey: /etc/istio/ingressgateway-certs/tls.key
    hosts:
    - "google.com"
    - "httpbin.example.com"
```

`--set global.proxy.image=istio/proxyv2:1.1.2` does NOT have this issues (using 1.1.4 for all other components)
`--set global.proxy.image=istio/proxyv2:1.1.3` DOES have this issue
`--set global.proxy.image=istio/proxyv2:1.1.4` DOES have this issue

Root cause seems to be this PR: https://github.com/envoyproxy/envoy/pull/6022. This effect seems to be an unintended consequence of that PR.

@costinm @silentdai @duderino

**Comment by PiotrSikora** [comment]
@howardjohn could you paste config that's pushed to Envoy? I believe it should show up in k8s logs when Envoy is logging at the `trace` level.

**Comment by silentdai** [comment]
I am wonder if a proper fix is to reshape the generated config at pilot.
I don't have the log to confirm but I highly suspect that there will be two filterchains of which
Port-DesitationPort-DestinationIP are the same, while beneath that tree node,
the first filter chain add "httpbin.example.com" node first, and then the second filter chain try to add 2 nodes with "google.com" and "httpbin.example.com". The later "httpbin.example.com" hit this throw

https://github.com/envoyproxy/envoy/blob/master/source/server/listener_manager_impl.cc#L428

**Comment by silentdai** [comment]
@howardjohn @costinm 
Let's suppose the envoy did not throw exceptions somehow,
If a connection is hitting httpbin.example.com, what would envoy choose to reach access httpbin? 
Another way to ask is what if secret is different(although the comment says it must be same)?

**Comment by howardjohn** [comment]
Filter chains are matched in order, which is the same order specified in the Gateway config

**Comment by howardjohn** [comment]
Tried to reproduce this with simpler set up1.1.4, default install
```yaml
apiVersion: networking.istio.io/v1alpha3
kind: VirtualService
metadata:
  name: httpbin
spec:
  hosts:
  - httpbin.example.com
  gateways:
  - mygateway
  tls:
  - match:
    - port: 443
      sni_hosts:
      - httpbin.example.com
    route:
    - destination:
        host: my-httpbin
        port:
          number: 443
---
apiVersion: networking.istio.io/v1alpha3
kind: Gateway
metadata:
  name: mygateway
spec:
  selector:
    istio: ingressgateway # use istio default ingress gateway
  servers:
  - port:
      number: 443
      name: https
      protocol: HTTPS
    tls:
      mode: PASSTHROUGH
    hosts:
    - httpbin.example.com

  - port:
      number: 443
      name: https-other
      protocol: HTTPS
    tls:
      mode: PASSTHROUGH
    hosts:
    - "google.com"
    - "httpbin.example.com"
```
Now I get `Error adding/updating listener(s) 0.0.0.0_443: error adding listener '0.0.0.0:443': multiple filter chains with the same matching rules are defined`

**Comment by PiotrSikora** [comment]
@howardjohn Filter chains are NOT matched in order, it's a decision tree.

**Comment by howardjohn** [comment]
@PiotrSikora thanks, you are right. I thought I read that recently. docs are at https://www.envoyproxy.io/docs/envoy/latest/api-v2/api/v2/listener/listener.proto#listener-filterchainmatch

**Comment by PiotrSikora** [comment]
This seems to be working as expected, the same host cannot be used in two otherwise equivalent filter chains, since the selection between the 2 would be impossible.

I don't think it was working before envoyproxy/envoy#6022 either.

Why do you think this is an issue?

**Comment by howardjohn** [comment]
Here is the config generated by istio-proxy 1.1.2

```
    {
     "version_info": "2019-04-30T00:19:53Z/2",
     "listener": {
      "name": "0.0.0.0_443",
      "address": {
       "socket_address": {
        "address": "0.0.0.0",
        "port_value": 443
       }
      },
      "filter_chains": [
       {
        "filter_chain_match": {
         "server_names": [
          "httpbin.example.com"
         ]
        },
        "tls_context": {
         "common_tls_context": {
          "tls_certificates": [
           {
            "certificate_chain": {
             "filename": "/etc/istio/ingressgateway-certs/tls.crt"
            },
            "private_key": {
             "filename": "/etc/istio/ingressgateway-certs/tls.key"
            }
           }
          ],
          "alpn_protocols": [
           "h2",
           "http/1.1"
          ]
         },
         "require_client_certificate": false
        },
        "filters": [
         {
          "name": "envoy.http_connection_manager",
          "config": {
           "tracing": {
            "overall_sampling": {
             "value": 100
            },
            "random_sampling": {
             "value": 1
            },
            "operation_name": "EGRESS",
            "client_sampling": {
             "value": 100
            }
           },
           "stat_prefix": "0.0.0.0_443",
           "rds": {
            "route_config_name": "https.443.https.mygateway.default",
            "config_source": {
             "ads": {}
            }
           },
           "use_remote_address": true,
           "set_current_client_cert_details": {
            "dns": true,
            "uri": true,
            "subject": true
           },
           "stream_idle_timeout": "0s",
           "forward_client_cert_details": "SANITIZE_SET",
           "access_log": [],
           "server_name": "istio-envoy",
           "http_filters": [
            {
             "name": "mixer",
             "config": {
              "service_configs": {
               "default": {
                "disable_check_calls": true
               }
              },
              "transport": {
               "check_cluster": "outbound|9091||istio-policy.istio-system.svc.cluster.local",
               "network_fail_policy": {
                "policy": "FAIL_CLOSE",
                "max_retry_wait": "1s",
                "base_retry_wait": "0.080s"
               },
               "report_cluster": "outbound|9091||istio-telemetry.istio-system.svc.cluster.local"
              },
              "default_destination_service": "default",
              "mixer_attributes": {
               "attributes": {
                "context.reporter.uid": {
                 "string_value": "kubernetes://istio-ingressgateway-6687d775b7-6mpww.istio-system"
                },
                "source.uid": {
                 "string_value": "kubernetes://istio-ingressgateway-6687d775b7-6mpww.istio-system"
                },
                "context.reporter.kind": {
                 "string_value": "outbound"
                },
                "source.namespace": {
                 "string_value": "istio-system"
                }
               }
              },
              "forward_attributes": {
               "attributes": {
                "source.uid": {
                 "string_value": "kubernetes://istio-ingressgateway-6687d775b7-6mpww.istio-system"
                }
               }
              }
             }
            },
            {
             "name": "envoy.cors"
            },
            {
             "name": "envoy.fault"
            },
            {
             "name": "envoy.router"
            }
           ],
           "generate_request_id": true,
           "upgrade_configs": [
            {
             "upgrade_type": "websocket"
            }
           ]
          }
         }
        ]
       },
       {
        "filter_chain_match": {
         "server_names": [
          "google.com",
          "httpbin.example.com"
         ]
        },
        "tls_context": {
         "common_tls_context": {
          "tls_certificates": [
           {
            "certificate_chain": {
             "filename": "/etc/istio/ingressgateway-certs/tls.crt"
            },
            "private_key": {
             "filename": "/etc/istio/ingressgateway-certs/tls.key"
            }
           }
          ],
          "alpn_protocols": [
           "h2",
           "http/1.1"
          ]
         },
         "require_client_certificate": false
        },
        "filters": [
         {
          "name": "envoy.http_connection_manager",
          "config": {
           "forward_client_cert_details": "SANITIZE_SET",
           "server_name": "istio-envoy",
           "access_log": [],
           "http_filters": [
            {
             "name": "mixer",
             "config": {
              "transport": {
               "report_cluster": "outbound|9091||istio-telemetry.istio-system.svc.cluster.local",
               "check_cluster": "outbound|9091||istio-policy.istio-system.svc.cluster.local",
               "network_fail_policy": {
                "policy": "FAIL_CLOSE",
                "max_retry_wait": "1s",
                "base_retry_wait": "0.080s"
               }
              },
              "service_configs": {
               "default": {
                "disable_check_calls": true
               }
              },
              "default_destination_service": "default",
              "mixer_attributes": {
               "attributes": {
                "source.uid": {
                 "string_value": "kubernetes://istio-ingressgateway-6687d775b7-6mpww.istio-system"
                },
                "context.reporter.kind": {
                 "string_value": "outbound"
                },
                "source.namespace": {
                 "string_value": "istio-system"
                },
                "context.reporter.uid": {
                 "string_value": "kubernetes://istio-ingressgateway-6687d775b7-6mpww.istio-system"
                }
               }
              },
              "forward_attributes": {
               "attributes": {
                "source.uid": {
                 "string_value": "kubernetes://istio-ingressgateway-6687d775b7-6mpww.istio-system"
                }
               }
              }
             }
            },
            {
             "name": "envoy.cors"
            },
            {
             "name": "envoy.fault"
            },
            {
             "name": "envoy.router"
            }
           ],
           "generate_request_id": true,
           "upgrade_configs": [
            {
             "upgrade_type": "websocket"
            }
           ],
           "tracing": {
            "overall_sampling": {
             "value": 100
            },
            "random_sampling": {
             "value": 1
            },
            "client_sampling": {
             "value": 100
            },
            "operation_name": "EGRESS"
           },
           "stat_prefix": "0.0.0.0_443",
           "use_remote_address": true,
           "rds": {
            "route_config_name": "https.443.https-other.mygateway.default",
            "config_source": {
             "ads": {}
            }
           },
           "set_current_client_cert_details": {
            "subject": true,
            "uri": true,
            "dns": true
           },
           "stream_idle_timeout": "0s"
          }
         }
        ]
       }
      ],
      "listener_filters": [
       {
        "name": "envoy.listener.tls_inspector"
       }
      ]
     },
     "last_updated": "2019-04-30T00:20:28.317Z"
    }
   ]
```

After my previous comment with the PASSTHROUGH gateway config, I think the following:

* Istio should reject these configs
* Envoy still has a bug. It should fail with `multiple filter chains with the same matching rules are defined` not the error we get - it just happens to be caught by that change.

**Comment by silentdai** [comment]
With extra changes we might make filter chain in order. However, is the order based match we must promise? 

Some alternative approach:
- Make it an istio side undefined behavior, so envoy is free to build the filter chain match tree as long as envoy doesn't throw exception.
- Reject user request. It is unknown to me if istio could detect this improper config when there is stale state.

**Comment by silentdai** [comment]
As we already broke the first match rule for long, we could explore options in parallel
- istio: Determine the what is the best the users. It's not that bad if istio wrongly rejected at small chance, but it is terrible if the spec pass the validation but lead to chaos. Should also considered if it is easy to sanitize the spec. I would be crazy if I submit a 10K line spec but the response is "not valid somewhere"
- envoy: Tolerate a config above config. The bottom line is that envoy should accept the config not at the risk of crashing. And if only chain A and chain B are conflicting, probably it's reasonable to pick either A or B. But not choose unrelated chain C. 

WDYT

**Comment by howardjohn** [comment]
Istio: we can pretty easily add validation here. Probably in https://github.com/istio/istio/blob/master/pilot/pkg/model/gateway.go.

Envoy: Either reject all or reject none. Right now we have a weird scenario where some stuff is rejected but some is not. Given Istio will validate, which one it doesn't doesn't matter too much?

**Comment by howardjohn** [comment]
By the way, not sure if this was clear, but this does not apply for HTTP. HTTP we don't create a fitler chain match, and instead match everything on the port and direct to the routes, where this is already handled properly

**Comment by howardjohn** [comment]
The reason the config I posted above with PASSTHROUGH is rejected in 1.1.2 is because I have no VirtualService for google.com, so it gets filtered out. Then the two filterChainMatches are *exactly* the same, not just that they have some overlap. A quick look at the code confirms the check is just an exact equality check.

**Comment by howardjohn** [comment]
This is fixed in 1.1.5, although we may want to do some followup work in Envoy
